rfd/0100-proxy-ssh-grpc.md

```diff
@@ -0,0 +1,197 @@
+service ConnectionProxyService {
+  // GetClusterDetails provides cluster information from the perspective of the Proxy
+  rpc GetClusterDetails(GetClusterDetailsRequest) returns (GetClusterDetailsResponse);
+  // DialNode establishes a connection to the target host
+  // 
+  // The bidirectional stream is a transport mechanism to send raw data between
+  // client and server. This allows other protocols(SSH) to be sent through the
+  // tunnel.
+  rpc DialNode(stream DialNodeRequest) returns (stream DialNodeResponse);
+}
+
+// Request for DialNode
+//
+// Connection flow
+// -> Target (client)
+// <- Connection (server)
+// <-> Frame/AgentFrame (client/server)
+// 
+// Separating the Frame and AgentFrame allows
+// SSH and the SSH Agent Protocol to be multiplexed
+// over the same connection.
+message DialNodeRequest {
+  oneof payload {
+    // Contains the information to identify and connect
+    // to the target host
+    Target dial_target = 1;
+    // Raw data to transmit
+    DataFrame frame = 2;
+  }
+}
+
+// Response for DialNodeHost
+message DialNodeResponse {
+  oneof payload {
+    // Cluster configuration details
+    ClusterDetails details = 1;
+    // Raw data to transmit
+    DataFrame frame = 2;
+  }
+}
+
+// Protocol represents the supported protocols
+// that can be proxied.
+enum Protocol {
+  PROTOCOL_UNSPECIFIED = 0;
+  // SSH
+  PROTOCOL_SSH = 1;
+  // SSH Agent
+  PROTOCOL_SSH_AGENT = 2;
+}
+
+// A data frame carrying the payload and 
+// procotol. Allows for multiplexing different
+// protocols on the same stream.
+message DataFrame {
+  // The protocol that sent the data
+  Protocol protocol = 1;
+  // The raw packet of data 
+  bytes payload = 2;
+}
+
+// Target indicates which server to connect to
+message Target {
+  // The server hostname/ip/uuid
+  string host = 1;
+  // The cluster the server is a member of
+  string cluster = 2;
+  // The OS login 
+  string login = 3;
+}
+
+// Request for GetClusterDetails.
+message GetClusterDetailsRequest {
+  // The cluster to get details for
+  string cluster = 1;
+}
+
+// Response for GetClusterDetails.
+message GetClusterDetailsResponse {
+  // Cluster configuration details
+  ClusterDetails details = 1;
+}
+
+// ClusterDetails contains details about the cluster configuration
+message ClusterDetails {
+  // If proxy recording mode is enabled
+  bool recording_proxy = 1;
+  // If the cluster is running in FIPS mode
+  bool fips_enabled = 2;
+}
+```
+
+The `DialNode` RPC establishes a connection to a Node on behalf of the user.
+There is an initial ceremony which must be completed by both sides of the stream
+before raw data can be transported. The client must first send a `Target`
+message which declares the target server that the connection is for. If the
+target exists and session control allows, the server will establish the
+connection and respond with a `ConnectionDetails` message.
+
+After the initial ceremony, either side of the connection may transmit `Frame`s
```

Comment:
We could save a roundtrip if the server could send its version string together with the `ConnectionDetails`. Perhaps `DialNodeResponse` could just hold both messages, and it'd just be illegal for the server to send a non-null `ConnectionDetails` after its first message?

The same could be done for `DialNodeRequest`, but it would be useless for the SSH protocol.

Reply:
With the SSH protocol it wouldn't be unexpected to have data (the `SSH-2.0-Teleport` line) available to send immediately after connecting; if we allowed the ConnectionDetails message together with a data frame we could send both in a single grpc stream message, which could save a TCP roundtrip and some overhead.